File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.py
class Node:
    def __init__(self, value):
        self.value = value
        self.parent = self
        self.size = 1

class UnionFind:
    def find(self, node1):
        if node1 == node1.parent:
            return node1
        return self.find(node1.parent)

    def union(self, node1, node2):
        parent1 = self.find(node1)
        parent2 = self.find(node2)

        if parent1 != parent2:
            parent2.parent = parent1
            parent1.size += parent2.size
        return parent1.size

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:

        hashmap = {}
        uf = UnionFind()
        maxSize, size = 0, 0

        for num in nums:
            if num not in hashmap:
                hashmap[num] = Node(num)
                size = 1
                if num-1 in hashmap:
                    size = uf.union(hashmap[num-1], hashmap[num])
                if num+1 in hashmap:
                    size = uf.union(hashmap[num], hashmap[num+1])

            maxSize = max(maxSize, size)

        return maxSize

File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.java
class UnionFind {
    int[] parent;
    int[] size;

    public UnionFind(int[] nums){
        this.parent = new int[nums.length];
        this.size = new int[nums.length];

        for(int i=0; i < nums.length; i++){
            this.parent[i] = i;
            this.size[i] = 1;
        }
    }

    public int find(int num){
        if(this.parent[num] != num){
            this.parent[num] = find(this.parent[num]);
        }
        return this.parent[num];
    }

    public int union(int num1, int num2){
        int parent1 = find(num1);
        int parent2 = find(num2);

        if(parent1 != parent2){
            this.parent[parent2] = parent1;
            this.size[parent1] += this.size[parent2];
        }
        return size[parent1];
    }
}

class Solution {
    public int longestConsecutive(int[] nums) {

        int maxLength = 0;
        UnionFind uf = new UnionFind(nums);

        Map<Integer, Integer> map = new HashMap<>();

        for(int i=0; i<nums.length; i++){
            if(!map.containsKey(nums[i])){
                 int currLength = 0;               
                map.put(nums[i], i);
                currLength = uf.size[i];
                if(map.containsKey(nums[i]-1)){
                    currLength = uf.union(map.get(nums[i]-1), i);
                }
                if(map.containsKey(nums[i]+1)){
                    currLength = uf.union(i, map.get(nums[i]+1));
                }            
                maxLength = Math.max(maxLength, currLength);
            }
        }

        return maxLength;
    }
}